Adding _from_bytes helpers as a foil for _to_bytes.

dhermes opened the pull request:
@nathanielmanistaatgoogle This is the child of #157.

I labeled it don't merge because I essentially just looked around for `.decode` and tried to preserve a lot of the stuff from the other PR, but haven't written any new tests.

Let's discuss (and I'll iterate) before merging.

**UPDATE**: I should note that I rebased the original PR if anyone wants to have a look: https://github.com/dhermes/oauth2client/tree/pr-157-rebased

Fixes #239 (~~but no tests yet~~ tested by `AssertionCredentialsTests.test_refresh_success_bytes`)

dhermes committed:
Adding _from_bytes helpers as a foil for _to_bytes.

dhermes commented:
I should also FYI to @saaros that this is happening

nathanielmanistaatgoogle commented on oauth2client/client.py:
```diff
@@ -279,8 +280,7 @@ def new_from_json(cls, s):
     """
-    if isinstance(s, bytes):
-      s = s.decode('utf-8')
+    s = _from_bytes(s)
```
Might it be worth using a different name for the result of _from_bytes, since it is potentially of a different type?

dhermes replied:
The old code was using `s = s.decode('utf-8')`, so I was just being mindless. Will change to `s_bytes` (and use a better varname than `s`).

New commit or fold into original?

nathanielmanistaatgoogle replied:
New commit is fine; thanks for asking.

nathanielmanistaatgoogle commented on oauth2client/client.py:
```diff
@@ -279,8 +280,7 @@ def new_from_json(cls, s):
       An instance of the subclass of Credentials that was serialized with
       to_json().
```
Should the "Args" section of this doc string be changed to better describe the input type as a byte string?

dhermes replied:
byte string or unicode string. `_from_bytes` will take something of `six.text_type` and just leave it as-is.

nathanielmanistaatgoogle replied:
In that case, let's be explicit about supporting both? Documentation that just says "string" suggests to the Python 3 reader that we're not describing the value completely.

dhermes replied:
Yup. I updated the whole docstring (first line was too long).

nathanielmanistaatgoogle commented:
Looks mostly good; what else do you want to do here?

dhermes committed:
Updating Credentials.new_from_json.

New _from_bytes behavior warrants update to docstring. Also
updated 1-letter or other non-descriptive variable names.

dhermes commented:
@nathanielmanistaatgoogle Sent the new commit.

I was looking for places that might need testing but kind of realized that the whole reason we test `_from_bytes` with all inputs is so that we don't have to send each of those types through every single code path which hits `_from_bytes`.

Also note I massively updated the `Credentials.new_from_json` method but left `s` as the only argument. This is because it is part of the public interface, so people can execute

```
Credentials.new_from_json(s=file_obj.read())
```

nathanielmanistaatgoogle commented:
Looks good to me - anything remaining to be done here before merge? Anyone else who should review this?

dhermes commented:
No and no. Merging.
